File: dungeons-1.21.11/src/main/java/fishmod/utils/Misc.java
package fishmod.utils;

import fishmod.utils.config.values.ExtraOptions;
import fishmod.utils.debug.Debug;
import config.practical.data.SoundData;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.hud.ChatHud;
import net.minecraft.client.gui.hud.InGameHud;
import net.minecraft.client.network.ClientPlayNetworkHandler;
import net.minecraft.client.network.ClientPlayerEntity;
import net.minecraft.entity.Entity;
import net.minecraft.sound.SoundEvent;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;

public class Misc {
    public static final MinecraftClient INSTANCE = MinecraftClient.getInstance();
    private static final Text ON = Text.literal("ON").formatted(Formatting.GREEN);
    private static final Text OFF = Text.literal("OFF").formatted(Formatting.RED);

    public static Vec3d getPos(Entity entity, double tickProgress) {
        double x = MathHelper.lerp(tickProgress, entity.lastRenderX, entity.getX());
        double y = MathHelper.lerp(tickProgress, entity.lastRenderY, entity.getY());
        double z = MathHelper.lerp(tickProgress, entity.lastRenderZ, entity.getZ());
        return new Vec3d(x, y, z);
    }

    public static double getDistance(Entity e1, Entity e2) {
        return getDistance(e1.getX(), e1.getZ(), e2.getX(), e2.getZ());
    }

    public static double getDistance(double x1, double z1, double x2, double z2) {
        return ((x1 - x2) * (x1 - x2)) + ((z1 - z2) * (z1 - z2));
    }

    public static void addChatMessage(Text text) {
        try {
            if (INSTANCE == null) return;
            InGameHud gameHud = INSTANCE.inGameHud;
            ChatHud hud = gameHud.getChatHud();
            forceMainThread(() -> hud.addMessage(Text.literal(ExtraOptions.textPrefix).append(text)));
        } catch (IndexOutOfBoundsException ignored) {
            Debug.LOGGER.error("Chat message failed to get added");
        }
    }

    public static Text getStatusText(boolean status) {
        return status ? ON : OFF;
    }

    public static void setTitle(Text text) {
        forceMainThread(() -> INSTANCE.inGameHud.setTitle(text));
    }

    public static void forceTitle(Text title, Text subtitle) {
        forceMainThread(() -> {
            INSTANCE.inGameHud.setTitle(title);
            INSTANCE.inGameHud.setSubtitle(subtitle);
        });
    }

    public static void executeCommand(String string) {
        ClientPlayNetworkHandler networkHandler = INSTANCE.getNetworkHandler();
        if (networkHandler == null) return;
        // sendChatCommand() expects no leading slash — strip one if the caller typed the command
        // the way they'd type it in chat.
        String trimmed = string.trim();
        String command = trimmed.startsWith("/") ? trimmed.substring(1) : trimmed;
        forceMainThread(() -> networkHandler.sendChatCommand(command));
    }

    public static void sendSound(SoundEvent soundEvent, float volume, float pitch) {
        ClientPlayerEntity player = INSTANCE.player;
        if (player == null) return;
        forceMainThread(() -> player.playSound(soundEvent, volume, pitch));
    }

    public static void sendSound(SoundData soundData) {
        sendSound(SoundEvent.of(soundData.getSound()), soundData.getVolume(), soundData.getPitch());
    }

    public static void forceMainThread(Runnable runnable) {
        if (INSTANCE.isOnThread()) {
            runnable.run();
        } else {
            INSTANCE.executeSync(runnable);
        }
    }
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/utils/Misc.kt
package fishmod.utils

import config.practical.data.SoundData
import fishmod.utils.config.values.ExtraOptions
import fishmod.utils.debug.Debug
import net.minecraft.client.MinecraftClient
import net.minecraft.entity.Entity
import net.minecraft.sound.SoundEvent
import net.minecraft.text.Text
import net.minecraft.util.Formatting
import net.minecraft.util.math.MathHelper
import net.minecraft.util.math.Vec3d

object Misc {
    @JvmField val INSTANCE: MinecraftClient = MinecraftClient.getInstance()
    private val ON: Text = Text.literal("ON").formatted(Formatting.GREEN)
    private val OFF: Text = Text.literal("OFF").formatted(Formatting.RED)

    @JvmStatic
    fun getPos(entity: Entity, tickProgress: Double): Vec3d {
        val x = MathHelper.lerp(tickProgress, entity.lastRenderX, entity.x)
        val y = MathHelper.lerp(tickProgress, entity.lastRenderY, entity.y)
        val z = MathHelper.lerp(tickProgress, entity.lastRenderZ, entity.z)
        return Vec3d(x, y, z)
    }

    @JvmStatic
    fun getDistance(e1: Entity, e2: Entity): Double {
        return getDistance(e1.x, e1.z, e2.x, e2.z)
    }

    @JvmStatic
    fun getDistance(x1: Double, z1: Double, x2: Double, z2: Double): Double {
        return ((x1 - x2) * (x1 - x2)) + ((z1 - z2) * (z1 - z2))
    }

    @JvmStatic
    fun addChatMessage(text: Text) {
        try {
            val instance = INSTANCE ?: return
            val gameHud = instance.inGameHud
            val hud = gameHud.chatHud
            forceMainThread { hud.addMessage(Text.literal(ExtraOptions.textPrefix).append(text)) }
        } catch (ignored: IndexOutOfBoundsException) {
            Debug.LOGGER.error("Chat message failed to get added")
        }
    }

    @JvmStatic
    fun getStatusText(status: Boolean): Text {
        return if (status) ON else OFF
    }

    @JvmStatic
    fun setTitle(text: Text) {
        forceMainThread { INSTANCE.inGameHud.setTitle(text) }
    }

    @JvmStatic
    fun forceTitle(title: Text, subtitle: Text) {
        forceMainThread {
            INSTANCE.inGameHud.setTitle(title)
            INSTANCE.inGameHud.setSubtitle(subtitle)
        }
    }

    @JvmStatic
    fun executeCommand(string: String) {
        val networkHandler = INSTANCE.networkHandler ?: return
        // sendChatCommand() expects no leading slash — strip one if the caller typed the command
        // the way they'd type it in chat.
        val trimmed = string.trim()
        val command = if (trimmed.startsWith("/")) trimmed.substring(1) else trimmed
        forceMainThread { networkHandler.sendChatCommand(command) }
    }

    @JvmStatic
    fun sendSound(soundEvent: SoundEvent, volume: Float, pitch: Float) {
        val player = INSTANCE.player ?: return
        forceMainThread { player.playSound(soundEvent, volume, pitch) }
    }

    @JvmStatic
    fun sendSound(soundData: SoundData) {
        sendSound(SoundEvent.of(soundData.sound), soundData.volume, soundData.pitch)
    }

    @JvmStatic
    fun forceMainThread(runnable: Runnable) {
        if (INSTANCE.isOnThread) {
            runnable.run()
        } else {
            INSTANCE.executeSync(runnable)
        }
    }
}
